torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1110,6 +1110,121 @@ void IndexSwizzle::handle(Swizzle2D* swizzle_2d) {
   IndexCompute::handle(swizzle_2d);
 }
 
+namespace {
+
+//! Check if the index of a parallel loop should be subsituted with
+//! zero.
+//!
+//! Zero substitution only happens with the BID parallel types with
+//! Local Or Shared tensors or the TID parallel types with Local
+//! tensors.
+//!
+//! This check is straightforward for consumers, but for producers
+//! the substitution is only done when the producer uses the same
+//! parallel type as the loop parallel type.
+//!
+//! If there's a mapped producer IterDoamin and that ID is
+//! parallelized, there are a couple of cases depending on the
+//! parallel type and the producer memory type:
+//!
+//! Loop PT, producer PT, producer mem -> index
+//! - BID, TID/Serial, Shared / Local -> use BID
```

Comment:
In this case, the producer ID is either TID or serial, but the mapped consumer (or loop) ID is BID parallel. More concretely:

```
tv0: [I0]
tv1 = tv0
tv2 = tv1

tv1->axis(0)->parallelize(ParallelType::Serial);
tv2->axis(0)->parallelize(ParallelType::BIDx);
```

This would look like:

```
for (i) {
  tv1[i] = tv0[...];
}
tv2[blockIdx.x] = tv1[...];
```

The question is what the index of `tv1` as the producer of `tv2` should look like. Suppose `tv1` is on shared memory or local memory, usually we don't use `blockIdx.x` to index shared memory as it's not shared across a grid and thus just use zero instead, but in this case indexing `tv1` as a producer requires to use `blockIdx.x`.

Memory promotion is done before this point, i.e., at the time of parallel validation (see class `SyncMap`).